restapi/auth.go

```diff
@@ -11,38 +14,99 @@ import (
 	"github.com/uwblueprint/shoe-project/restapi/rest"
-	"golang.org/x/crypto/bcrypt"
-	"gorm.io/gorm"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
 )
 
+var oauthconfig *oauth2.Config
+
+func generateStateOauthCookie(w http.ResponseWriter) (string, error) {
+	duration, err := config.GetTokenExpiryDuration()
+	if err != nil {
+		return "", err
+	}
+	var expiration = time.Now().Add(duration)
+
+	b := make([]byte, 16)
+	rand.Read(b)
+	state := base64.URLEncoding.EncodeToString(b)
+	fmt.Println(state)
+	cookie := http.Cookie{Name: "oauthstate", Value: state, Expires: expiration, HttpOnly: true} // in prod, set Secure to true
+	http.SetCookie(w, &cookie)
+
+	return state, nil
+}
+
 func (api api) Login(w http.ResponseWriter, r *http.Request) render.Renderer {
-	var user models.User
-	err := json.NewDecoder(r.Body).Decode(&user)
+	oauthconfig = &oauth2.Config{
+		ClientID:     config.GetGoogleClientId(),
+		ClientSecret: config.GetGoogleClientSecret(),
+		RedirectURL:  "http://localhost:8900/api/auth/callback",
+		Scopes: []string{
+			"https://www.googleapis.com/auth/userinfo.email",
+		},
+		Endpoint: google.Endpoint,
+	}
+
+	state, err := generateStateOauthCookie(w)
+	if err != nil {
+		return rest.ErrInternal(api.logger, err)
+	}
+
+	url := oauthconfig.AuthCodeURL(state)
+	http.Redirect(w, r, url, http.StatusTemporaryRedirect)
+
+	return rest.JSONStatusOK(url)
+}
+
+func (api api) AuthCallback(w http.ResponseWriter, r *http.Request) render.Renderer {
+	expectedState, _ := r.Cookie("oauthstate")
+	if expectedState.Value != r.FormValue("state") {
+		return rest.ErrUnauthorized("Invalid oauth state")
+	}
+
+	// exchange received authorization code for an access token
+	token, err := oauthconfig.Exchange(context.Background(), r.FormValue("code"))
 	if err != nil {
-		rest.ErrInvalidRequest(api.logger, "Invalid login payload", err)
+		return rest.ErrUnauthorized("Invalid authorization code")
 	}
 
-	// Find user with username
-	var dbUser models.User
-	if err = api.database.Where("username=?", user.Username).First(&dbUser).Error; err != nil {
-		if err == gorm.ErrRecordNotFound {
-			return rest.ErrNotFound(fmt.Sprintf("Could not find user with username %s", user.Username))
-		}
+	// use access token to create a client which we use to access user info
+	client := oauthconfig.Client(context.Background(), token)
+	response, err := client.Get("https://www.googleapis.com/oauth2/v3/userinfo")
+	if err != nil {
+		return rest.JSONStatusOK("Invalid access token")
+	}
+	defer response.Body.Close()
+	user := models.User{}
+	err = json.NewDecoder(response.Body).Decode(&user)
+	if err != nil {
 		return rest.ErrInternal(api.logger, err)
 	}
 
-	// Validate password
-	if err = bcrypt.CompareHashAndPassword([]byte(dbUser.Password), []byte(user.Password)); err != nil {
-		return rest.ErrUnauthorized(fmt.Sprintf("Wrong password for username %s", user.Username))
+	// redirect if invalid
+	if user.Hd != "uwblueprint.org" && user.Hd != "theshoeproject.online" {
```

Comment:
Should we make this a struct method on user?

```go
func (user u) hasValidEmail() bool {
    return user.Hd == "uwblueprint.org" || user.Hd == "theshoeproject.online"
}
```

that way if we decide to change `Hd` in the future, we don't have to reach into the API

Reply:
i actually think it's the opposite...if things change in the future, we don't want to reach into database-layer logic?

Reply:
Well I actually think @AbhiPrasad suggestion makes sense. Since this logic is custom to us (i.e has nothing to do with the actual google OAUTH flow) it would be better to make it a member of the User model to encapsulate that logic.

Reply:
Ok, I've added this logic in a hook before any user record is created (see latest commit)